pkg/R/RDD.R

```diff
@@ -1323,6 +1323,88 @@ setMethod("zipWithIndex",
            lapplyPartitionsWithIndex(x, partitionFunc)
          })
 
+#' Zip an RDD with another RDD.
+#'
+#' Zips this RDD with another one, returning key-value pairs with the
+#' first element in each RDD second element in each RDD, etc. Assumes
+#' that the two RDDs have the same number of partitions and the same
+#' number of elements in each partition (e.g. one was made through
+#' a map on the other).
+#'
+#' @param x An RDD to be zipped.
+#' @param other Another RDD to be zipped.
+#' @return An RDD zipped from the two RDDs.
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' rdd1 <- parallelize(sc, 0:4)
+#' rdd2 <- parallelize(sc, 1000:1004)
+#' collect(zipRDD(rdd1, rdd2))
+#' # list(list(0, 1000), list(1, 1001), list(2, 1002), list(3, 1003), list(4, 1004))
+#'}
+#' @rdname zipRDD
+#' @aliases zipRDD,RDD
+setMethod("zipRDD",
+          signature(x = "RDD", other = "RDD"),
+          function(x, other) {
+            n1 <- numPartitions(x)
+            n2 <- numPartitions(other)
+            if (n1 != n2) {
+              stop("Can only zip RDDs which have the same number of partitions.")
+            }
+
+            # Check if corresponding partitions of both RDDs have the same
+            # number of elements.
+            getPartitionLengthFunc <- function(part) {
+              list(length(part))
+            }
+            partitionLengths1 <- collect(lapplyPartition(x, getPartitionLengthFunc))
```

Comment:
In order to call JavaRDD.zip(), we need to make sure that both `x` and `y` have the same serialization mode and the same number of items in each partitions (you can have multiple Array[Byte] in one partition after coalesce()). So we always need to call `reserialize`.

If we want to avoid to introduce another serialization mode (it's not too bad actually), we need a way to tell which item come from `x` or `y`. so we could do something like this:
```
append_length <- function(items) { items[[length(items) + 1]] <- length(items) + 1; items }
x <- lapplyPartitions(x, append_length)
y <- lapplyPartitions(y, append_length)
jrdd <- callJMethod(getJRDD(x), "zip", getJRDD(y))
r <- RDD(jrdd)
lapplyPartitions(r, function(items) {
  l <- length(items)
  length_y <- items[[l]]
  length_x <- items[[l - length_y]]
  stopifnot(l == length_x + length_y)
  if (length_x != length_y) {
    stop("can not zip")
  }
  lapply(1:(length_x-1), function(i) {
    list(items[[i]], items[[length_x + i]])
  })
```